src/resolv/lookup/srv.rs

```diff
@@ -0,0 +1,215 @@
+//! Looking up SRV records.
+
+use futures::{Async, Future, Poll};
+use std::net::IpAddr;
+use ::bits::{DName, DNameBuf, DNameSlice, MessageBuf, ParsedDName, ParseResult};
+use ::iana::{Rtype, Class};
+use ::rdata::{Srv, A, Aaaa};
+use super::super::{Query, Resolver};
+use super::super::error::{Error, Result};
+use super::search::SearchIter;
+use rand;
+use rand::distributions::{IndependentSample, Range};
+
+
+//------------ lookup_records ------------------------------------------------
+
+/// Creates a future that looks up SRV records.
+///
+/// TODO: documentation
+pub fn lookup_srv<N>(resolver: Resolver, name: N) -> LookupSRV
+                  where N: AsRef<DNameSlice> {
+    let name = name.as_ref();
+    let mut search = SearchIter::new(resolver.clone(), name);
+    let search_name = search.as_mut().map(|s| s.next().unwrap());
+    let query_name = match search_name {
+        Some(ref name) => name,
+        None => name
+    };
+    let query = resolver.clone().query((query_name, Rtype::Srv, Class::In));
+    LookupSRV {
+        resolver: resolver,
+        query: query,
+        search: search,
+    }
+}
+
+
+//------------ LookupRecords -------------------------------------------------
+
+/// The future returned by [`lookup_srv()`].
+///
+/// [`lookup_srv()`]: fn.lookup_srv.html
+pub struct LookupSRV {
+    /// The resolver to run queries on.
+    resolver: Resolver,
+
+    /// The current querry.
+    query: Query,
+
+    /// An optional search list iterator for searching a name.
+    search: Option<SearchIter>,
+}
+
+
+//--- Future
+
+impl Future for LookupSRV {
+    type Item = FoundSRVs;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let err = match self.query.poll() {
+            Ok(Async::NotReady) => return Ok(Async::NotReady),
+            Ok(Async::Ready(item)) => return Ok(Async::Ready(FoundSRVs::new(item)?)),
+            Err(err) => err
+        };
+        let name = match self.search {
+            None => return Err(err),
+            Some(ref mut search) => {
+                match search.next() {
+                    None => return Err(err),
+                    Some(name) => name,
+                }
+            }
+        };
+        self.query = self.resolver.clone()
+                         .query((name, Rtype::Srv, Class::In));
+        self.poll()
+    }
+}
+
+
+//------------ FoundSRVs ----------------------------------------------------
+
+/// The value returned by a successful SRV lookup.
+///
+/// TODO: doc
+///
+/// The `canonical_name()` method returns the canonical name of the host for
+/// which the addresses were found.
+#[derive(Clone, Debug)]
+pub struct FoundSRVs {
+    /// The canonical domain name for the host.
+    canonical: DNameBuf,
+
+    /// All the IP addresses we’ve got.
+    results: Vec<SRVItem>
+}
+
+#[derive(Clone, Debug)]
+pub struct SRVItem {
+    host: DNameBuf,
+    port: u16,
+    addrs: Option<Vec<IpAddr>>
```

Comment:
That is kind of what I'm doing now, SrvItem holds a state which is either `Unresolved(name)` or `Resolved(found_hosts)`.